{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GATK Variant Calling Pipelines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# telepítések\n",
    "\n",
    "\n",
    "git clone https://github.com/broadinstitute/picard.git\n",
    "cd picard/\n",
    "./gradlew shadowJar\n",
    "\n",
    "cd ..\n",
    "git clone https://github.com/broadinstitute/gatk.git   \n",
    "cd gatk\n",
    "./gradlew installAll  \n",
    "\n",
    "cd ..\n",
    "wget https://github.com/RealTimeGenomics/rtg-tools/releases/download/3.9/rtg-tools-3.9-linux-x64.zip\n",
    "unzip rtg-tools-3.9-linux-x64.zip\n",
    "\n",
    "\n",
    "#https://cloud.google.com/sdk/docs/\n",
    "wget https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-194.0.0-linux-x86_64.tar.gz\n",
    "\n",
    "tar xzf google-cloud-sdk-194.0.0-linux-x86_64.tar.gz\n",
    "\n",
    "./google-cloud-sdk/install.sh\n",
    "./google-cloud-sdk/bin/gcloud init\n",
    "\n",
    "# https://hail.is/ érdemes lenne megnézni, hogy mi is ez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adatbázisok\n",
    "\n",
    "http://cancer.sanger.ac.uk/cosmic\n",
    "    \n",
    "# Genome Aggregation Database (gnomAD)\n",
    "http://gnomad.broadinstitute.org/downloads\n",
    "# gsutil -m cp -r gs://gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr20.vcf.bgz gnomad_data\n",
    "\n",
    "# Panel of Normals (PoN)\n",
    "\n",
    "# germline population frequency = G Pop Freqs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "export PATH=${PATH}:/home/sn/Downloads/install/src/gatk\n",
    "\n",
    "PICARD=\"/home/sn/Downloads/install/src/picard/build/libs/picard.jar\"\n",
    "\n",
    "export PATH=${PATH}:/home/sn/Downloads/install/rtg-tools-3.9\n",
    "\n",
    "# /home/sn/Downloads/install/IGV_2.4.10/igv.sh\n",
    "\n",
    "cd /home/sn/Desktop/GATK_Vcall/wd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Germline Variant Calling\n",
    "\n",
    "![title](https://us.v-cdn.net/5019796/uploads/editor/mz/tzm69d8e2spl.png)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Germline Variants\n",
    "\n",
    "java -jar $PICARD CreateSequenceDictionary R=ref.fasta O=ref.dict\n",
    "\n",
    "samtools faidx ref.fasta\n",
    "\n",
    "samtools index mother.bam mother.bai\n",
    "samtools index father.bam father.bai\n",
    "samtools index son.bam son.bai\n",
    "\n",
    "# a htslib (samtools) telepítéssel jönnek\n",
    "bgzip -c motherGIABsnps.vcf > motherGIABsnps.vcf.gz\n",
    "tabix -p vcf motherGIABsnps.vcf.gz\n",
    "\n",
    "\n",
    "gatk BaseRecalibrator \\\n",
    "  -R ref.fasta \\\n",
    "  -I mother.bam \\\n",
    "  --known-sites motherGIABsnps.vcf.gz \\\n",
    "  -O 1st_recal.table\n",
    "\n",
    "  \n",
    "gatk ApplyBQSR \\\n",
    "  -R ref.fasta \\\n",
    "  -I mother.bam \\\n",
    "  -bqsr 1st_recal.table \\\n",
    "  -O mother_bqsr.bam\n",
    "  \n",
    "    \n",
    "gatk BaseRecalibrator \\\n",
    "  -R ref.fasta \\\n",
    "  -I mother_bqsr.bam \\\n",
    "  --known-sites motherGIABsnps.vcf.gz \\\n",
    "  -O 2nd_recal.table\n",
    "\n",
    "\n",
    "gatk AnalyzeCovariates \\\n",
    "  -before 1st_recal.table \\\n",
    "  -after 2nd_recal.table \\\n",
    "  -plots plots.pdf\n",
    "\n",
    "\n",
    "gatk HaplotypeCaller \\\n",
    "  -R ref.fasta \\\n",
    "  -I mother.bam \\\n",
    "  -O mother01.vcf \\\n",
    "  -L 20:10,000,000-10,200,000\n",
    "  \n",
    "  \n",
    "# realigned reads\n",
    "gatk HaplotypeCaller \\\n",
    "  -R ref.fasta \\\n",
    "  -I mother.bam \\\n",
    "  -O mother01debug.vcf \\\n",
    "  -bamout mother01debug.bam \\\n",
    "  -L 20:10,002,371-10,002,546 -ip 100\n",
    "\n",
    "        \n",
    "# Variant calling in gVCF mode\n",
    "gatk HaplotypeCaller \\\n",
    "  -R ref.fasta \\\n",
    "  -I mother.bam \\\n",
    "  -O mother.g.vcf \\\n",
    "  -ERC GVCF \\\n",
    "  -L 20:10,000,000-10,200,000\n",
    "  \n",
    "\n",
    "gatk HaplotypeCaller \\\n",
    "  -R ref.fasta \\\n",
    "  -I son.bam \\\n",
    "  -O son.g.vcf \\\n",
    "  -bamout son01debug.bam \\\n",
    "  -ERC GVCF \\\n",
    "  -L 20:10,000,000-10,200,000\n",
    "\n",
    "    \n",
    "gatk HaplotypeCaller \\\n",
    "  -R ref.fasta \\\n",
    "  -I father.bam \\\n",
    "  -O father.g.vcf \\\n",
    "  -bamout father01debug.bam \\\n",
    "  -ERC GVCF \\\n",
    "  -L 20:10,000,000-10,200,000\n",
    "\n",
    "###########################################################################  \n",
    "# GVCF-based joint calling\n",
    "###########################################################################\n",
    "# https://software.broadinstitute.org/gatk/documentation/article.php?id=3893  \n",
    "\n",
    "# Consolidate GVCFs using GenomicsDBImport\n",
    "gatk GenomicsDBImport \\\n",
    "  -V mother.g.vcf \\\n",
    "  -V father.g.vcf \\\n",
    "  -V son.g.vcf \\\n",
    "  --genomicsdb-workspace-path csalad \\\n",
    "  --intervals 20:10,000,000-10,200,000\n",
    "\n",
    "# Joint call\n",
    "# Joint Genotyping\n",
    "gatk GenotypeGVCFs \\\n",
    "  -R ref.fasta \\\n",
    "  -V gendb://csalad \\\n",
    "  -O csaladGGVCF.vcf \\\n",
    "  -L 20:10,000,000-10,200,000 \n",
    "\n",
    "        \n",
    "###########################################################################  \n",
    "# Filter Variants\n",
    "###########################################################################\n",
    "\n",
    "# Variant Recalibration\n",
    "gatk VariantRecalibrator \\\n",
    "  -R human.fasta \\\n",
    "  -V raw.SNPs.vcf \\\n",
    "  -resource [tags]:filename.vcf \\\n",
    "  -an DP -an QD -an FS -an MQRankSum {...} \\\n",
    "  -mode SNP \\\n",
    "  -recal-file raw.SNPs.recal \\\n",
    "  -tranches-file raw.SNPs.tranches \\\n",
    "  -rscript-file recal.plots.R\n",
    "\n",
    "\n",
    "# ApplyVQSR applies the filtering threshold\n",
    "gatk ApplyVQSR \\\n",
    "  -R human.fasta \\\n",
    "  -V raw.vcf \\\n",
    "  -mode SNP \\\n",
    "  -recal-file raw.SNPs.recal \\\n",
    "  -tranches-file raw.SNPs.tranches \\\n",
    "  -O recal.SNPs.vcf \\\n",
    "  -ts-filter-level 99.0\n",
    "\n",
    "\n",
    "################################################ megnézni: Convolutional Neural Networks    \n",
    "\n",
    "###########################################################################  \n",
    "# Refine Genotypes\n",
    "###########################################################################\n",
    "\n",
    "# Genotype call\n",
    "gatk CalculateGenotypePosteriors \\\n",
    "  -R reference.fasta \\\n",
    "  -V input.vcf \\\n",
    "  -ped family.ped \\\n",
    "  -supporting population.vcf \\\n",
    "  -O output.vcf\n",
    "\n",
    "# Variants with Posterior Qualities\n",
    "gatk VariantFiltration \\\n",
    "  -R reference.fasta \\\n",
    "  -V input.vcf \\\n",
    "  --filter-expression “GQ<20” \\\n",
    "  --filter-name “lowGQ” \\\n",
    "  -O output.vcf    \n",
    "    \n",
    "# Annotate Variants\n",
    "gatk VariantAnnotator \\\n",
    "  -R reference.fasta \\\n",
    "  -V input.vcf \\\n",
    "  -A PossibleDeNovo \\\n",
    "  -O output.vcf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Somatic Variant Calling\n",
    "\n",
    "![title](https://us.v-cdn.net/5019796/uploads/editor/nk/6t0uqznajjmg.png)\n",
    "\n",
    "https://software.broadinstitute.org/gatk/documentation/article?id=11136\n",
    "\n",
    "https://gatkforums.broadinstitute.org/dsde/discussion/11136/how-to-call-somatic-mutations-using-gatk4-mutect2/p1?new=1\n",
    "\n",
    "https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php\n",
    "\n",
    "https://github.com/gatk-workflows/gatk4-somatic-snvs-indels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /media/sn/oldsys/gatk_bundle/3-somatic\n",
    "\n",
    "# T-N Pair\n",
    "gatk Mutect2 \\\n",
    "  -R ref/Homo_sapiens_assembly38.fasta \\\n",
    "  -I bams/tumor.bam \\\n",
    "  -I bams/normal.bam \\\n",
    "  -tumor HCC1143_tumor \\\n",
    "  -normal HCC1143_normal \\\n",
    "  -pon resources/chr17_m2pon.vcf.gz \\\n",
    "  --germline-resource resources/chr17_af-only-gnomad_grch38.vcf.gz \\\n",
    "  --af-of-alleles-not-in-resource 0.0000025 \\\n",
    "  --disable-read-filter MateOnSameContigOrNoMappedMateReadFilter \\\n",
    "  -L resources/chr17plus.interval_list \\\n",
    "  -O sandbox/1_somatic_m2.vcf.gz \\\n",
    "  -bamout sandbox/2_tumor_normal_m2.bam\n",
    "\n",
    "# The Panel of Normals (PoN) has a vital role that fills a gap between the matched normal and the population resource. \n",
    "# Mutect2 uses the PoN (i) to catch additional sites of noise in sequencing # data, like mapping artifacts or other \n",
    "# somewhat random but systematic artifacts of sequencing, and data processing and (ii) to filter sites of common \n",
    "# germline variation, especially for cases where a common germline variants resource is unavailable. \n",
    "# Ideally, the PoN should include samples that were sequenced on the same platform using the same chemistry and \n",
    "# analyzed using the same reference genome and workflow (technically similar). However, even an unmatched PoN is \n",
    "# better than no PoN at all. This is because mapping artifacts and polymerase slippage errors occur for pretty\n",
    "# much the same genomic loci for short read sequencing approaches. \n",
    "\n",
    "# To make your own PoN, run Mutect2 in \"tumor-only mode\" on each normal BAM individually then \n",
    "# run CreateSomaticPanelOfNormals on the resulting VCFs.\n",
    "\n",
    "#Run this command on each normal sample that will be included in the PoN\n",
    "gatk Mutect2 \\\n",
    "  -R ref/Homo_sapiens_assembly38.fasta \\\n",
    "  -I bams/HG00190.bam \\\n",
    "  -tumor HG00190 \\\n",
    "  --disable-read-filter MateOnSameContigOrNoMappedMateReadFilter \\\n",
    "  -L resources/chr17plus.interval_list \\\n",
    "  -O sandbox/3_HG00190.vcf.gz\n",
    "\n",
    "\n",
    "# Collate all the normal VCFs into a single callset with CreateSomaticPanelOfNormals\n",
    "gatk CreateSomaticPanelOfNormals \\\n",
    "  -vcfs sandbox/3_HG00190.vcf.gz \\\n",
    "  -vcfs sandbox/4_NA19771.vcf.gz \\\n",
    "  -vcfs sandbox/5_HG02759.vcf.gz \\\n",
    "  -O sandbox/6_threesamplepon.vcf.gz\n",
    "\n",
    "\n",
    "# cross-sample contamination\n",
    "gatk GetPileupSummaries \\\n",
    "  -I bams/tumor.bam \\\n",
    "  -V resources/chr17_small_exac_common_3_grch38.vcf.gz \\\n",
    "  -O sandbox/7_tumor_getpileupsummaries.table\n",
    "\n",
    "\n",
    "# Estimate contamination with CalculateContamination\n",
    "gatk CalculateContamination \\\n",
    "  -I sandbox/7_tumor_getpileupsummaries.table \\\n",
    "  -O sandbox/8_tumor_calculatecontamination.table\n",
    "\n",
    "\n",
    "# filtering on annotations and contamination\n",
    "gatk FilterMutectCalls \\\n",
    "  -V sandbox/1_somatic_m2.vcf.gz \\\n",
    "  --contamination-table sandbox/8_tumor_calculatecontamination.table \\\n",
    "  -O sandbox/9_somatic_oncefiltered.vcf.gz\n",
    "\n",
    "\n",
    "#CollectSequencingArtifactMetrics (Picard)\n",
    "#https://software.broadinstitute.org/gatk/documentation/tooldocs/current/picard_analysis_artifacts_CollectSequencingArtifactMetrics.php\n",
    "java -jar $PICARD CollectSequencingArtifactMetrics \\\n",
    "  I= bams/tumor.bam \\\n",
    "  O= sandbox/tumor.preadapter_detail_metrics \\\n",
    "  R= ref/Homo_sapiens_assembly38.fasta\n",
    "    \n",
    "\n",
    "# optionally filter by orientation bias\n",
    "gatk FilterByOrientationBias \\\n",
    "  -V sandbox/9_somatic_oncefiltered.vcf.gz \\\n",
    "  --artifact-modes 'G/T' \\\n",
    "  -P sandbox/tumor.preadapter_detail_metrics \\\n",
    "  -O sandbox/10_tumor_oxog_twicefiltered.vcf.gz\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![title](https://us.v-cdn.net/5019796/uploads/editor/x2/rzka6s5hie5r.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Somatic Variant Calling with Tumor Samples Only\n",
    "\n",
    "https://gatkforums.broadinstitute.org/gatk/discussion/6469/somatic-variant-calling-with-tumor-samples-only\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# VarDict\n",
    "\n",
    "https://www.nature.com/articles/srep43169?WT.feed_name=subjects_biotechnology\n",
    "https://academic.oup.com/nar/article/44/11/e108/2468301\n",
    "\n",
    "https://github.com/AstraZeneca-NGS/VarDict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "git clone https://github.com/AstraZeneca-NGS/VarDict.git\n",
    "\n",
    "export PATH=${PATH}:/home/sn/Downloads/install/src/VarDict\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# R\n",
    "\n",
    "http://bioconductor.org/packages/release/bioc/vignettes/VariantTools/inst/doc/tutorial.pdf\n",
    "\n",
    "http://bioconductor.org/packages/release/bioc/html/gmapR.html\n",
    "\n",
    "https://cran.r-project.org/web/packages/vcfR/vignettes/intro_to_vcfR.html\n",
    "\n",
    "https://www.bioconductor.org/packages/release/BiocViews.html#___SomaticMutation\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tumor only\n",
    "\n",
    "Evaluating somatic tumor mutation detection without matched normal samples\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5584341/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
